app/javascript/lib/scp-api.jsx

```diff
@@ -243,6 +242,33 @@ export function setupRenewalForReadOnlyToken(studyAccession) {
   }, renewalTime)
 }
 
+
+/**
+ * Renew userAccessToken, very similiar to above function to renew readOnlyToken but
+ * this is for the userAccessToken instead
+ * setting of the token on the window is done in the generate function in the user model
+ */
+export function setUpRenewalForUserAccessToken() {
+  let renewalTime = 60000 * 55 * 1000 // 55 minutes default
+
+  // if the token object has been set already use that expiration time instead
+  if (window.SCP.userAccessTokenObject) {
+    const userAccessTokenObject = camelcaseKeys(window.SCP.userAccessTokenObject)
+    // Set the renewal time to be 5 minutes before expiration
+    renewalTime = userAccessTokenObject.expiresIn * 1000 - FIVE_MINUTES
+  }
+
+  setTimeout(async () => {
+    const apiUrl = `/site/renew_user_access_token`
+    const [response] = await scpApi(apiUrl, defaultInit())
+    const userAccessTokenObj = response
+    window.SCP.userAccessTokenObject = userAccessTokenObj
```

Comment:
This seems to be the only place `window.SCP.userAccessTokenObject` is initialized.  That entails this API call will be issued at page load, on every page load, for signed-in users.  Such load seems unnecessary.

Both `readOnlyAccessTokenObject` and the original `window.SCP.userAccessToken` are set on page load, server-side, in a Rails template.  The latter is set like so in [line 36 of /app/views/layouts/application.html.erb](https://github.com/broadinstitute/single_cell_portal_core/blob/850f7da6db0aa961087392a5df0af5e79e69b562/app/views/layouts/application.html.erb#L36):

```
window.SCP.userAccessToken = '<%= get_user_access_token(current_user) %>';
```

Note how the analogous read-only access token is initialized around [line 5 in /app/views/site/_study_visualize.html.erb](https://github.com/broadinstitute/single_cell_portal_core/blob/850f7da6db0aa961087392a5df0af5e79e69b562/app/views/site/_study_visualize.html.erb#L5):

```
    if (typeof window.SCP.readOnlyToken === 'undefined') {
      window.SCP.readOnlyTokenObject = <%= raw get_read_access_token(@study, current_user).to_json %>
      window.SCP.readOnlyToken = window.SCP.readOnlyTokenObject.access_token
    }
```

Let's apply that pattern in line 36 of `/app/views/layouts/application.html.erb`.  

Also, importantly, note how in `setupRenewalForReadOnlyToken` we do:

```
    const [response] = await scpApi(apiUrl, defaultInit())
    const readOnlyTokenObject = response
    window.SCP.readOnlyToken = readOnlyTokenObject.accessToken
```

We need to similarly set `window.SCP.userAccessToken` in this function.  That's [used](https://github.com/search?q=repo%3Abroadinstitute%2Fsingle_cell_portal_core%20userAccessToken&type=code) in a few critical places, which are ultimately used to set the `Bearer` token in e.g. `scp-api.jsx`, `metrics-api.jsx`, and elsewhere.

Reply:
ah okay yeah I'll make the update in application.html.erb to set up the tokenObject

For `window.SCP.userAccessToken` it is set in the `generate_user_access_token()` function that is called to get the token in scp-api so it shouldn't need to be set again here. I mentioned that in the comment right below this line, but if that doesn't seem clear I can look at moving things around. Since `generate_user_access_token()` was an existing function I didn't want to change it up unnecessarily which is why I didn't pull that out, but I agree it's not necessarily as clear.

Reply:
I was curious about that comment, i.e.:

> `// the userAccessToken is set in the window in the generate_access_token in the user model`

A bit more specifically, `userAccessToken` is set in the `window` in the `application` view via `generate_access_token`, which is defined in the `user` model.

> For window.SCP.userAccessToken it is set in the generate_user_access_token() function that is called to get the token in scp-api so it shouldn't need to be set again here.

Right now, as you allude to, `userAccessToken` is only set in `application.html.erb`.  The problem I'm drawing attention to here is that when `userAccessToken` expires while a page is in view, the current PR's code will not renew it, even though the function says it will in its docstring.  

The problem the `renewFooToken` functions solve is that access tokens expire in 60 minutes after being granted, but a page can be viewed (i.e., not reloaded) for longer and the user expects its features to still work at, say, minute 75.  This is described more fully in #1747.  

More concretely, the problem is that it's `userAccessToken` which is used as the bearer token for client-side access to e.g. the SCP API and Bard API (via `getAccessToken()`, which resolves to `window.SCP.userAccessToken`), but `userAccessToken` is not getting its value renewed here.  The variable `userAccessTokenObject` is not directly used for the bearer token (nor should it).  To fix the expiry-during-page-view issue this function is intended to fix, the `window.SCP.userAccessToken` needs to be assigned the renewed access token in `setUpRenewalForUserAccessToken`.

Reply:
(To keep track of blocking issues, I updated my original comment in this thread to mark this as a place I think a change needs to be made.)